agenta-backend/agenta_backend/routers/observability_router.py

```diff
@@ -0,0 +1,206 @@
+async def get_spans_of_trace(
+    trace_id: str,
+    stoken_session: SessionContainer = Depends(verify_session()),
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    spans = await get_trace_spans(trace_id, **kwargs)
+    print("Spans: ", spans)
+    return spans
+
+
+@router.get("/costs/{trace_id}/", response_model=float)
+async def get_costs_of_trace(
+    trace_id: str,
+    stoken_session: SessionContainer = Depends(verify_session()),
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    costs = await costs_of_llm_run(trace_id, **kwargs)
+    return costs
+
+
+@router.get("/tokens/{trace_id}/", response_model=int)
+async def get_tokens_of_trace(
+    trace_id: str,
+    stoken_session: SessionContainer = Depends(verify_session()),
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    tokens = await tokens_of_llm_run(trace_id, **kwargs)
+    return tokens
+
+
+@router.get("/latency/{trace_id}/", response_model=float)
+async def get_latency_of_trace(
+    trace_id: str,
+    stoken_session: SessionContainer = Depends(verify_session()),
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    latency = await latency_of_llm_run(trace_id, **kwargs)
+    return latency
+
+
+@router.get("/inputs/{trace_id}/", response_model=TraceInputs)
+async def get_trace_inputs(
+    trace_id: str, stoken_session: SessionContainer = Depends(verify_session())
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    inputs = await inputs_of_llm_run(trace_id, **kwargs)
+    return inputs
+
+
+@router.get("/outputs/{trace_id}/", response_model=TraceOutputs)
+async def get_trace_outputs(
+    trace_id: str, stoken_session: SessionContainer = Depends(verify_session())
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    outputs = await outputs_of_llm_run(trace_id, **kwargs)
+    return outputs
+
+
+@router.put("/traces/{trace_id}/", response_model=bool)
+async def update_trace(
+    trace_id: str,
+    payload: UpdateTrace,
+    stoken_session: SessionContainer = Depends(verify_session()),
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    trace = await update_trace_status(trace_id, payload, **kwargs)
+    return trace
+
+
+@router.post("/feedbacks/{trace_id}/", response_model=Feedback)
+async def create_feedback(
+    trace_id: str,
+    payload: CreateFeedback,
+    stoken_session: SessionContainer = Depends(verify_session()),
+):
+    # Get user and org id
+    kwargs: dict = await get_user_and_org_id(stoken_session)
+    feedback = await add_feedback_to_trace(trace_id, payload, **kwargs)
+    return feedback
+
+
+@router.get("/feedbacks/{trace_id}/", response_model=List[Feedback])
+async def get_feedbacks(
```

Comment:
I think this might be a bit of an issue.
Let's say you'd like to have a dashboard where you show all the traces with a column showing the average feedback. Or have one where you compare the feedback for two variants.
Currently this might be very expensive, no?
You need to list all the traces for a variant, then go for each and list all the feedbacks, aggregate these, and show them.


Does it make sense to have feedback a subcollection in the trace collection? Can you please look at the pros and cons for this?